source/fuzz/fuzzer_pass_permute_function_variables.cpp

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#include "source/fuzz/fuzzer_pass_permute_function_variables.h"
+
+namespace spvtools {
+namespace fuzz {
+
+FuzzerPassPermuteFunctionVariables::FuzzerPassPermuteFunctionVariables(
+    opt::IRContext* ir_context, TransformationContext* transformation_context,
+    FuzzerContext* fuzzer_context,
+    protobufs::TransformationSequence* transformations)
+    : FuzzerPass(ir_context, transformation_context, fuzzer_context,
+                 transformations) {} // here call parent constructor 
+
+// use default destructor 
+FuzzerPassPermuteFunctionVariables::~FuzzerPassPermuteFunctionVariables() = default;
+
+void FuzzerPassPermuteFunctionVariables::Apply() {
+for (const auto& function : *GetIRContext()->module()) {
+    uint32_t function_id = function.result_id();
+
+    // enty point mean something like e.g. main(), so skip it. 
+    // because FunctionIsEntryPoint Returns |true| if one of entry points has function id |function_id|
+    if (fuzzerutil::FunctionIsEntryPoint(GetIRContext(), function_id)) {
+      continue;
+    }
+
+    if (!GetFuzzerContext()->ChoosePercentage(
+            GetFuzzerContext()->GetChanceOfPermuteFunctionVariables())) {
+      continue;
+    }
+
+
+
+    ApplyTransformation(TransformationSwapFunctionVariables(,,function_id));
```

Comment:
This does not seem correct.

I suggest you approach it as follows.

First, determine the number of local variable declarations that the function has - i.e., the number of OpVariable instructions it contains. Refer to the SPIR-V specification for details of where OpVariable instructions may appear in a function.

If there are N local variable declarations, decide on a random permutation of the set [0, 1, ..., N-1]. Look at FuzzerPassPermuteParameters for an example of how to get a random permutation.

Once you have a random permutation of the form:

```
0  1  2  ...  N-1
i0 i1 i2 ... iN-1
```

then figure out how to express this as a product of *transpositions* - i.e., swaps. This page might help:

http://mathonline.wikidot.com/decomposition-of-permutations-as-products-of-transpositions

and there are many other resources online that explain how to do it (but do ask if you need help as it's a bit tricky).

Once you know that your permutation can be expressed by e.g. swapping v1 with v2, then swapping v2 with v4, then swapping v4 with v1 (or whatever), call ApplyTransformation as many times as needed, in each case telling the transformation which variables should be swapped.

Reply:
Thanks, Dr. @afd  I understood the product of transposition, but I wonder if the permutation is a set of random products of transpositions or it follows a certain formula like that I've found on the internet: 
[(a1,a2,...,as)=(as,as−1)∘(as,as−2)∘...∘(as,a2)∘(as,a1)](http://mathonline.wikidot.com/decomposition-of-permutations-as-products-of-transpositions)  
what do you think?

Reply:
A permutation can always be written as a product of cycles, e.g.

(1, 2, 3)(4, 6)(7, 9, 10)

is a permutation on the set {1, ..., 10} (it leaves 5 fixed).

Then each cycle can be decomposed into transpositions in the manner you describe above.

Reply:
yes, Dr. @afd, I've understand this and use this concept in my code.

Reply:
Excellent.